engine/loader/den.5
.\"	$Csoft: den.5,v 1.2 2004/05/12 05:33:33 vedge Exp $
.\"
.\" Copyright (c) 2003, 2004, 2005 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd JUNE 19, 2003
.Dt DEN 5
.ds vT Agar API Reference
.ds oS Agar 1.0
.Os
.Sh NAME
.Nm den
.Nd agar den archive format
.Sh DESCRIPTION
The
.Nm
file format has been designed to encapsulate data (such as graphics or audio),
with support for language-specific overrides and searching facilities.
.Pp
Integers are stored in network format, strings are length-encoded.
A
.Nm
archive starts with the
.Sq agar den
string, followed by a 64-bit version code and
a
.Nm
header, which looks like:
.Bd -literal
struct den {
	char	 hint[DEN_HINT_MAX];	/* Hint on type of content */
	char	 name[DEN_NAME_MAX];	/* Name of the package */
	char	*author;		/* Author of the contents */
	char	*copyright;		/* Copyright notice */
	char	*descr;			/* Description of contents */
	char	*keywords;		/* Comma-separated keywords */

	struct den_member *members;
	Uint32		  nmembers;
};
.Ed
.Pp
The
.Va hint
field gives a hint about the type of contents, and is only for searching
purposes.
The
.Va name
is an identifier for the whole archive.
.Pp
The
.Va author ,
.Va copyright ,
.Va descr
are optional.
.Va keywords
is provided for searching purposes.
.Pp
The
.Va members
table describes the contents of the archive.
Each element, in its encoded form, looks like:
.Bd -literal
struct den_member {
	Uint32	 name_len;
	char	 name[DEN_MEMBER_NAME_MAX];	/* Identifier */
	Uint32	 lang_len;
	char	 lang[DEN_MEMBER_LANG_MAX];	/* ISO 639 language code */
	Uint32	 offs;				/* Position in archive */
	Uint32	 size;				/* Size in bytes */
};
.Ed
.Pp
The
.Va name
string identifies this archive member.
.Va lang
is an ISO 639 language code, or a null string to indicate that this member
contains no language-specific data.
The loader chooses which member to read depending on language preferences.
.Sh SEE ALSO
.Xr agar 3
.Sh HISTORY
The
.Nm
file format first appeared in Agar 1.0


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

engine/loader/den.c
/*	$Csoft: den.c,v 1.6 2005/01/05 04:44:04 vedge Exp $	*/

/*
 * Copyright (c) 2003, 2004, 2005 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <compat/strlcpy.h>

#include <engine/error/error.h>

#include <sys/types.h>
#include <SDL_types.h>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>

#include <engine/loader/netbuf.h>
#include <engine/loader/version.h>
#include <engine/loader/integral.h>
#include <engine/loader/string.h>
#include <engine/loader/den.h>

const struct version den_ver = {
	"agar den",
	0, 0
};

/* Read a den header and its mappings. */
static int
den_read_header(struct den *den)
{
	Uint32 i;

	copy_string(den->hint, den->buf, sizeof(den->hint));
	copy_string(den->name, den->buf, sizeof(den->name));

	den->author = read_string(den->buf);
	den->copyright = read_string(den->buf);
	den->descr = read_string(den->buf);
	den->keywords = read_string(den->buf);

	den->nmembers = read_uint32(den->buf);
	den->members = Malloc(den->nmembers*sizeof(struct den_member), M_LOADER);

	for (i = 0; i < den->nmembers; i++) {
		struct den_member *memb = &den->members[i];

		copy_string(memb->name, den->buf, sizeof(memb->name));
		copy_string(memb->lang, den->buf, sizeof(memb->lang));
		den->members[i].offs = (off_t)read_uint32(den->buf);
		den->members[i].size = (size_t)read_uint32(den->buf);
	}
	return (0);
}

/* Write a den header and skip the mapping table. */
void
den_write_header(struct den *den, int nmemb)
{
	Uint32 i;

	write_string(den->buf, den->hint);
	write_string(den->buf, den->name);

	write_string(den->buf, den->author);
	write_string(den->buf, den->copyright);
	write_string(den->buf, den->descr);
	write_string(den->buf, den->keywords);

	/* Initialize the mapping table. */
	den->members = Malloc(nmemb*sizeof(struct den_member), M_LOADER);
	den->nmembers = (Uint32)nmemb;
	for (i = 0; i < den->nmembers; i++) {
		struct den_member *memb = &den->members[i];

		memset(memb->name, 'B', sizeof(memb->name));
		memset(memb->lang, 'o', sizeof(memb->lang));
	}

	write_uint32(den->buf, den->nmembers);
	
	/* Skip the mappings. */
	den->mapoffs = netbuf_tell(den->buf);
	netbuf_seek(den->buf, den->nmembers*DEN_MAPPING_SIZE, SEEK_CUR);
}

/* Write the den mappings. */
void
den_write_mappings(struct den *den)
{
	Uint32 i;

	netbuf_seek(den->buf, den->mapoffs, SEEK_SET);

	for (i = 0; i < den->nmembers; i++) {
		struct den_member *memb = &den->members[i];

		write_uint32(den->buf, (Uint32)sizeof(memb->name));
		netbuf_write(memb->name, sizeof(memb->name), 1, den->buf);
		write_uint32(den->buf, (Uint32)sizeof(memb->lang));
		netbuf_write(memb->lang, sizeof(memb->lang), 1, den->buf);

		write_uint32(den->buf, (Uint32)memb->offs);
		write_uint32(den->buf, (Uint32)memb->size);
	}
}

/* Open a den archive; load the header as well in read mode. */
struct den *
den_open(const char *path, enum den_open_mode mode)
{
	struct den *den;

	den = Malloc(sizeof(struct den), M_LOADER);
	den->buf = netbuf_open(path, (mode == DEN_READ) ? "rb" : "wb",
	    NETBUF_BIG_ENDIAN);
	if (den->buf == NULL) {
		Free(den, M_LOADER);
		return (NULL);
	}

	switch (mode) {
	case DEN_READ:
		if (version_read(den->buf, &den_ver, NULL) == -1 ||
		    den_read_header(den) == -1) {
			goto fail;
		}
		break;
	case DEN_WRITE:
		version_write(den->buf, &den_ver);
		break;
	}
	return (den);
fail:
	netbuf_close(den->buf);
	Free(den, M_LOADER);
	return (NULL);
}

void
den_close(struct den *den)
{
	netbuf_close(den->buf);

	Free(den->author, 0);
	Free(den->copyright, 0);
	Free(den->descr, 0);
	Free(den->keywords, 0);
	Free(den->members, M_LOADER);
	Free(den, M_LOADER);
}

/* Import the contents of a file in a den archive. */
int
den_import_file(struct den *den, int ind, const char *name, const char *lang,
    const char *infile)
{
	char buf[8192];
	struct den_member *memb;
	FILE *f;
	size_t size, rrv;
	off_t offs;
	
	offs = netbuf_tell(den->buf);
	size = 0;

	if ((f = fopen(infile, "rb")) == NULL) {
		error_set("%s: %s", infile, strerror(errno));
		return (-1);
	}
	for (;;) {
		rrv = fread(buf, 1, sizeof(buf), f);
		size += rrv;

		netbuf_write(buf, rrv, 1, den->buf);
		if (rrv < sizeof(buf)) {
			if (feof(f)) {
				break;
			} else {
				error_set("%s: read error", infile);
				return (-1);
			}
		}
	}
	fclose(f);

	memb = &den->members[ind];
	strlcpy(memb->name, name, sizeof(memb->name));
	strlcpy(memb->lang, lang, sizeof(memb->lang));
	memb->offs = offs;
	memb->size = size;
	return (0);
}
